gadopt/inverse.py

```diff
@@ -0,0 +1,265 @@
+from firedrake import CheckpointFile
+import firedrake.utils
+from firedrake.adjoint import *  # noqa: F401
+from pathlib import Path
+from mpi4py import MPI
+from pyadjoint.optimization.optimization_solver import OptimizationSolver
+import pyadjoint.optimization.rol_solver as pyadjoint_rol
+import ROL
+
+# emulate the previous behaviour of firedrake_adjoint by automatically
+# starting the tape
+continue_annotation()
+
+_vector_registry = []
+
+
+class ROLSolver(pyadjoint_rol.ROLSolver):
+    def __init__(self, problem, parameters, inner_product="L2", vector_class=pyadjoint_rol.ROLVector, vector_args=[]):
+        OptimizationSolver.__init__(self, problem, parameters)
+        self.rolobjective = pyadjoint_rol.ROLObjective(problem.reduced_functional)
+        x = [p.tape_value() for p in self.problem.reduced_functional.controls]
+        self.rolvector = vector_class(x, *vector_args, inner_product=inner_product)
```

Comment:
Yeah, good point. It's because we need to override the class used for the vectors within the solver if we're using the ROL checkpointing. It might actually be sufficient to `super().__init__()` and then just re-initialise `self.rolvector`...